File: src/apps/daemon/internal/platform/system/traffic_meter_smoother_test.go
package system

import (
	"testing"
	"time"
)

func TestTrafficMeterSmoother(t *testing.T) {
	m := NewTrafficMeterSmoother()
	m.Start(1000, 2000)

	// Step 1: 1 second, +102400 rx, +51200 tx
	s1 := m.SampleNow(1000+102400, 2000+51200, 1*time.Second)
	if s1.ReceivedBytes != 102400 || s1.SentBytes != 51200 {
		t.Fatalf("unexpected totals: %+v", s1)
	}
	if s1.DownloadBytesPerSec != 102400 || s1.UploadBytesPerSec != 51200 {
		t.Fatalf("unexpected rates: %+v", s1)
	}

	// Step 2: 1 second, +102400 rx, 0 tx
	s2 := m.SampleNow(1000+204800, 2000+51200, 1*time.Second)
	if s2.DownloadBytesPerSec != 102400 {
		t.Fatalf("expected download rate to remain 102400, got %d", s2.DownloadBytesPerSec)
	}
	if s2.UploadBytesPerSec != 20480 {
		t.Fatalf("expected smoothed upload rate 20480, got %d", s2.UploadBytesPerSec)
	}
}

func TestFormatBytesAndRate(t *testing.T) {
	if FormatBytes(500) != "500 B" {
		t.Fatalf("unexpected: %s", FormatBytes(500))
	}
	if FormatBytes(1536) != "1.5 KB" {
		t.Fatalf("unexpected: %s", FormatBytes(1536))
	}
	if FormatBytes(10*1024*1024) != "10 MB" {
		t.Fatalf("unexpected: %s", FormatBytes(10*1024*1024))
	}
	if FormatRate(10*1024*1024) != "10 MB/s" {
		t.Fatalf("unexpected: %s", FormatRate(10*1024*1024))
	}
}

func TestSplitTunnelPolicy(t *testing.T) {
	p := SplitTunnelPolicy{
		Mode:     SplitModeOnly,
		Packages: []string{"com.example.browser", "com.luminet.vpn"},
	}

	eff := p.EffectivePackages("com.luminet.vpn")
	if len(eff) != 1 || eff[0] != "com.example.browser" {
		t.Fatalf("expected self exclusion, got %v", eff)
	}

	if p.ValidationError("com.luminet.vpn") != "" {
		t.Fatalf("expected no validation error, got %s", p.ValidationError("com.luminet.vpn"))
	}

	pEmpty := SplitTunnelPolicy{
		Mode:     SplitModeOnly,
		Packages: []string{"com.luminet.vpn"},
	}
	if pEmpty.ValidationError("com.luminet.vpn") == "" {
		t.Fatalf("expected validation error for empty only list")
	}
}

File: src/packages/lumicore/tests/traffic_meter_smoother_test.rs
use std::collections::HashSet;
use std::time::Duration;
use lumicore::diagnostics::{
    format_bytes, format_rate, SplitTunnelMode, SplitTunnelPolicy, TrafficMeterSmoother,
};

#[test]
fn test_traffic_meter_smoothing() {
    let mut meter = TrafficMeterSmoother::new();
    meter.start(1000, 2000);

    // After 1 second, rx climbed by 102400 (100 KB/s), tx by 51200 (50 KB/s)
    let s1 = meter.sample_now(1000 + 102400, 2000 + 51200, Duration::from_secs(1));
    assert_eq!(s1.received_bytes, 102400);
    assert_eq!(s1.sent_bytes, 51200);
    assert_eq!(s1.download_bytes_per_sec, 102400);
    assert_eq!(s1.upload_bytes_per_sec, 51200);

    // After another second, another 102400 rx, but tx is 0
    // smoothed_down should stay 102400 (0.4 * 102400 + 0.6 * 102400 = 102400)
    // smoothed_up should decay (0.4 * 51200 + 0.6 * 0 = 20480)
    let s2 = meter.sample_now(1000 + 204800, 2000 + 51200, Duration::from_secs(1));
    assert_eq!(s2.download_bytes_per_sec, 102400);
    assert_eq!(s2.upload_bytes_per_sec, 20480);
}

#[test]
fn test_format_bytes_and_rate() {
    assert_eq!(format_bytes(500), "500 B");
    assert_eq!(format_bytes(1536), "1.5 KB");
    assert_eq!(format_bytes(10 * 1024 * 1024), "10 MB");
    assert_eq!(format_bytes(1500 * 1024 * 1024), "1.5 GB");

    assert_eq!(format_rate(10 * 1024 * 1024), "10 MB/s");
}

#[test]
fn test_split_tunnel_policy() {
    let mut targets = HashSet::new();
    targets.insert("com.example.browser".to_string());
    targets.insert("com.luminet.vpn".to_string());

    let policy = SplitTunnelPolicy {
        mode: SplitTunnelMode::Only,
        targets,
    };

    // Self package must be excluded
    let effective = policy.effective_targets("com.luminet.vpn");
    assert_eq!(effective.len(), 1);
    assert!(effective.contains("com.example.browser"));
    assert!(!effective.contains("com.luminet.vpn"));

    // Validation passes with 1 target
    assert!(policy.validation_error("com.luminet.vpn").is_none());

    // Validation fails if only target was self
    let mut self_only = HashSet::new();
    self_only.insert("com.luminet.vpn".to_string());
    let policy_empty = SplitTunnelPolicy {
        mode: SplitTunnelMode::Only,
        targets: self_only,
    };
    assert!(policy_empty.validation_error("com.luminet.vpn").is_some());
}
